crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,499 @@
+/// For multiline `__all__` definitions that include comments,
+/// the fix is marked as unsafe, as it can be hard to tell where the comments
+/// should be moved to when sorting the contents of `__all__`.
+#[violation]
+pub struct UnsortedDunderAll;
+
+impl Violation for UnsortedDunderAll {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__all__` is not alphabetically sorted")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Sort `__all__` alphabetically".to_string())
+    }
+}
+
+pub(crate) fn sort_dunder_all_assign(
+    checker: &mut Checker,
+    node: &ast::StmtAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAssign { value, targets, .. } = node;
+    let [ast::Expr::Name(ast::ExprName { id, .. })] = targets.as_slice() else {
+        return;
+    };
+    sort_dunder_all(checker, id, value, parent);
+}
+
+pub(crate) fn sort_dunder_all_augassign(
+    checker: &mut Checker,
+    node: &ast::StmtAugAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAugAssign {
+        value,
+        target,
+        op: ast::Operator::Add,
+        ..
+    } = node
+    else {
+        return;
+    };
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+        return;
+    };
+    sort_dunder_all(checker, id, value, parent);
+}
+
+pub(crate) fn sort_dunder_all_annassign(
+    checker: &mut Checker,
+    node: &ast::StmtAnnAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAnnAssign {
+        target,
+        value: Some(val),
+        ..
+    } = node
+    else {
+        return;
+    };
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+        return;
+    };
+    sort_dunder_all(checker, id, val, parent);
+}
+
+fn sort_dunder_all(checker: &mut Checker, target: &str, node: &ast::Expr, parent: &ast::Stmt) {
+    if target != "__all__" {
+        return;
+    }
+
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    let locator = checker.locator();
+
+    let Some(dunder_all_val) = DunderAllValue::from_expr(node, locator) else {
+        return;
+    };
+
+    let new_dunder_all =
+        match dunder_all_val.construct_sorted_all(locator, parent, checker.stylist()) {
+            SortedDunderAll::AlreadySorted => return,
+            SortedDunderAll::Sorted(value) => value,
+        };
+
+    let dunder_all_range = dunder_all_val.range();
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, dunder_all_range);
+
+    if let Some(new_dunder_all) = new_dunder_all {
+        let applicability = {
+            if dunder_all_val.multiline
+                && checker.indexer().comment_ranges().intersects(node.range())
+            {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            }
```

Comment:
I tend to think that using unsafe to protect comments is a bit of a misuse of the concept, which is meant to protect against possible breakages via changes in semantics. But, this question has come up before, and I don't know that we have a unified answer.

\cc @zanieb

Reply:
I feel like as a _user_ of ruff I'd probably be a little annoyed if ruff did something like https://github.com/astral-sh/ruff/pull/9474#issuecomment-1888091353 to my carefully designed `__all__` definition, and then claimed it was a "safe" fix 😄 but I'll defer to you and Zanie :)

Reply:
That's a fair point. We would likely need to mark all fixes as unsafe following that reasoning because:
* some fixes don't handle comments at all
* Many fixes don't produce nicely formatted code. You could use the same argument that we should use unsafe because ruff could mess up my carefully hand-formatted code. 

Let's see what @zanieb thinks

Reply:
I've pushed 24616f3a55e18640c92b548177842ba65021bd9c to mark the fix as always being safe, but I've kept a note in the docs stating that it might sometimes move comments to unexpected locations